### PR TITLE
DRILL-7045 UDF string_binary java.lang.IndexOutOfBoundsException

### PR DESCRIPTION
**Pull request by jcmcote**
fix bug in string_binary, it was not doing reallocIfNeeded

### CODE REVIEW

**Comment by arina-ielchiieva** [comment]
@KazydubB please review.

File: exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestJsonNanInf.java [hunk 1]
```diff
@@ -236,6 +236,22 @@ public void testConvertFromJsonFunction() throws Exception {
 
 
 
+  @Test
+  public void testLargeStringBinary() throws Exception {
+    String chunk = "0123456789";
+    String data = "";
+    for (int i = 0; i < 1000; i++) {
+      data += chunk;
```
Comment (KazydubB) [inline exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestJsonNanInf.java]: It is preferable to change such string concatenations with `StringBuilder#append(String)`. (`String data` --> `StringBuilder data`).

File: exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestJsonNanInf.java [hunk 2]
```diff
@@ -236,6 +236,22 @@ public void testConvertFromJsonFunction() throws Exception {
...
 
+  @Test
+  public void testLargeStringBinary() throws Exception {
+    String chunk = "0123456789";
+    StringBuilder builder = new StringBuilder();
+    for (int i = 0; i < 1000; i++) {
+      builder.append(chunk);
+    }
+    String data = builder.toString();
+    String query = String.format("select string_binary(binary_string('%s')) from (values(1))", data);
```
Comment (KazydubB) [inline exec/java-exec/src/test/java/org/apache/drill/exec/vector/complex/writer/TestJsonNanInf.java]: Please change 
```
String query = String.format("select string_binary(binary_string('%s')) from (values(1))", data);
List<QueryDataBatch> results = testSqlWithResults(query);
RecordBatchLoader batchLoader = new RecordBatchLoader(getAllocator());
QueryDataBatch batch = results.get(0);
batch.release();
batchLoader.clear();
```
to 
```
test("select string_binary(binary_string('%s')) from (values(1))", data);
```

**Comment by sohami** [comment]
@jcmcote  - can you please address last comment from @KazydubB

**Comment by sohami** [comment]
@jcmcote - I have address @KazydubB comment in this commit and rebased on latest apache. Can you please make the change or pull in this commit so that we can close this PR ? https://github.com/sohami/drill/commit/7aaef8691a4a594442464301035ea3aefd7497dd

**Comment by arina-ielchiieva** [comment]
@sohami since there is not response from @jcmcote, you can merge two commits (first initial commit to preserve author and then your commit with changes after code review). By the way, thanks for doing this!

**Comment by sohami** [comment]
@arina-ielchiieva - Thanks for the suggestion. I will close this PR and open a new one with 2 commits and request @KazydubB to do a final review for it. Let me know if that sound's good.

**Comment by arina-ielchiieva** [comment]
Sure, that will be great.

**Pull request closed without merging**